### PR TITLE
UI: Add functions to check for and request macOS permissions

### PR DESCRIPTION
**Pull request by PatTheMav**
### Description
Adds functions to check and/or request specific macOS permissions (audio device access, video device access, accessibility access, and screen capture access).

By default only audio capture, video capture, and accessibility are requested on launch - the first two have straight-forward "Yes/No" prompts, the latter requires people to enable OBS in the settings application (but is required for hotkey functionality, independent
of scene setups).

### Motivation and Context
Reduce friction to gain necessary permissions on macOS, especially for fixed hotkey functionality.

### How Has This Been Tested?
Yes.

### Types of changes
- New feature (non-breaking change which adds functionality)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

File: UI/obs-app.cpp [hunk 1]
```diff
@@ -2070,6 +2070,13 @@ static int run_program(fstream &logFile, int argc, char *argv[])
 
 #if __APPLE__
 	InstallNSApplicationSubclass();
+	bool audio_permission =
+		CheckPermissionWithPrompt(kAudioDeviceAccess, true);
+	bool video_permission =
+		CheckPermissionWithPrompt(kVideoDeviceAccess, true);
+	bool accessibility_permission =
+		CheckPermissionWithPrompt(kAccessibility, true);
```
Comment (gxalpha) [inline UI/obs-app.cpp:2078]: Should these use the RequestPermission macro?

**Comment by gxalpha** [comment]
In general, I'm not sure if we should really just request these permissions on startup - especially the prompts for accessibility and screen capture have the tendency to just disappear behind the main OBS window (meaning users don't see them). Also users might wonder why OBS wants all of this even if they haven't done anything yet.
I wonder if we should have some kind of window saying "OBS needs permission x for this, y for this - which of these do you want to grant?". For reference, this is the style I'd think of (but for permissions instead, and this one is a bit too wordy):
image

**Review by gxalpha — COMMENTED** [review]
Noted this off-thread already but just so that this doesn't get lost: Now that we use an enum the blog calls aren't correct anymore (they're giving wrong information). Also I think it'd be great if each check always calls blog, not just when a prompt is requested.

**Comment by PatTheMav** [comment]
> Noted this off-thread already but just so that this doesn't get lost: Now that we use an enum the blog calls aren't correct anymore (they're giving wrong information). Also I think it'd be great if each check always calls blog, not just when a prompt is requested.

Right, it's only fixed for `kAccessibility` - waiting for your force push before making any other changes.

**Comment by gxalpha** [comment]
I didn't plan to do any more force pushes, did everything I wanted to in the one. Sorry for the confusion.

**Review by gxalpha — APPROVED** [review]
Works fine with the dialog (which currently is WIP). No issues on my side.

File: UI/obs-app.cpp [hunk 2]
```diff
@@ -2110,6 +2110,22 @@ static int run_program(fstream &logFile, int argc, char *argv[])
 		OBSTranslator translator;
 		program.installTranslator(&translator);
 
+		if (!created_log) {
+			create_log_file(logFile);
+			created_log = true;
+		}
+
+#ifdef __APPLE__
```
Comment (WizardCM) [inline UI/obs-app.cpp:2118]: Does this actually need to occur before `if (!already_running)`?

Comment (gxalpha) [reply UI/obs-app.cpp:2118]: We could probably move it further down if that's desirable

Comment (PatTheMav) [reply UI/obs-app.cpp:2118]: Good point, moved it accordingly.

File: UI/platform-osx.mm [hunk 3]
```diff
@@ -238,6 +242,138 @@ - (void)drawRect:(NSRect)dirtyRect
...
+						      (CFDictionaryRef)options)
+						      ? kPermissionAuthorized
+						      : kPermissionDenied);
+		}
+
+		blog(LOG_INFO, "[macOS] Permission for accessibility %s.",
+		     permissionResponse == kPermissionAuthorized ? "granted"
+								 : "denied");
+		return permissionResponse;
+		break;
```
Comment (WizardCM) [inline UI/platform-osx.mm:372]: Is this standard in Objective C and/or consistent with our other uses of switch statements?

Comment (PatTheMav) [reply UI/platform-osx.mm]: Removed the early returns, as every block assigns a correct value to `permissionresponse`.
